### PR TITLE
Switch from Progrock to OpenTelemetry

### PR DESCRIPTION
**Pull request by vito**
## Out with Progrock, in with OpenTelemetry

Now that OpenTelemetry has support for streaming logs, we are able to switch from Progrock and embrace an open standard and ecosystem instead.

This change should be mostly invisible to the user, unless they want to integrate with OpenTelemetry. In that case, they can set the [standard `OTEL_` environment variables](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#otlp-exporter) client-side, and all telemetry will be sent to the configured OTLP exporter, _without requiring any engine-side configuration_.

How does it work?

1. We use a custom OpenTelemetry pipeline that is able to send "live" spans to a configured exporter; this is how the TUI (and soon Cloud) is able to show you what's _happening_ rather than what _happened_.
2. To get telemetry data out of the engine and to the TUI, the engine implements an OpenTelemetry publish/subscribe pattern. The CLI subscribes to its own trace, and the engine publishes telemetry to the subscriber.
3. We originally needed this pub/sub just so the TUI can show engine activity, but since we already had to do the work anyway, we also re-process data received from the pub/sub and send _finished_ spans to your configured exporter. This way you only need to configure `OTEL_` on the client side!
4. DagQL instrumentation now creates an Otel span for each field resolution and adds base64-encoded `callpbv1.Call` as a span attribute. Not that this is a single `Call` and not a full `DAG`, which saves us ~10x (unpredictable) data sent over the wire, but implies the consumer needs to keep track of all `Call`s that it sees in order to grok inter-`Call` references.
5. Dagger Cloud will become an OTLP endpoint that is capable of receiving "live" spans.

That covers tracing. What about logging?

1. The Go OpenTelemetry SDK does not yet implement support for logging, so we have a package of our own, basically copy-pasta from the tracing stack with changes to support logging, since it's already supported everywhere else. This can be removed once https://github.com/open-telemetry/opentelemetry-go/pull/4955 lands.
2. Same as above, we have a pub/sub stack for logs, too. There's no "live" vs "complete" distinction for logs though.
3. Our shim detects the current trace and span, which Buildkit already propagates for us, and pipes the command's stdout/stderr to `io.Writer`s that send the logs to OpenTelemetry. This should maybe be upstreamed to Buildkit once 4955 lands.
4. DagQL instrumentation automatically configures `iotctx.Stdout` and `ioctx.Stderr` to send to the resolver's span.
5. To replace Progrock's "global logging" feature, we can now send logs with a special annotation that we can detect client-side to show them globally, while still maintaining a relationship to an originating span. This part isn't fully baked yet, but seems viable.
6. Dagger Cloud will become an OTLP endpoint that is capable of receiving logs for spans.
7. Engine logs can also be sent to Dagger Cloud as before, but I need to resurrect this code and confirm it works.

### Breaking changes
* The old TUIs have been removed. So both the `dagger run` DAG TUI, and the ancient interactive tree view TUI. These were just not worth translating to the new world, so this was finally the time to GC them. Apologies to the handful of users that still used the tree view style UI.
* The `--focus` flag is gone.
* The `pipeline()` APIs are now no-ops.

## New features / improvements

* The CLI now supports `curl` style verbosity flags.
	* `-v` keeps things from disappearing
	* `-vv` reveals internal/encapsulated spans
	* `-vvv` reveals spans that evaluated very quickly
* `dagger watch` allows you to see all activity across the engine, which is pretty neat. I've already used it for troubleshooting integration tests. It works by just subscribing to "everything" via the pub/sub service mentioned above.
* A span can be given a special attribute that tells the UI to collapse its children by default: `telemetry.Encapsulate()`.
* We now proactively shut down services on client close.
* We should no longer have to worry about missing the 'last few bits' of telemetry. The pub/sub system now waits for active spans to complete before detaching. So hopefully that means no more "runaway builds" or services left running forever.
* The TUI now supports a `Ctrl+\` keybind and `SIGQUIT`s itself, so if there's a hang for whatever reason, now we can more easily get a stack dump.
* The Engine now uses the same OpenTelemetry stack as the rest of Dagger, so hopefully we don't have to worry too much about dependency hell. (Knock on wood.)
* All the labels that used to be sent to Cloud as part of Pipelines are now associated to the CLI's OpenTelemetry resource instead.
* User-land code can directly hook in to the broader trace; we automatically configure `OTEL_` as appropriate in the shim.

## Regressions

* We no longer use Buildkit's progress stream, so we are at the mercy of their Otel integration, which is not as rich.
	* We no longer show `CACHED` status for Buildkit cache hits. :( Instead you'll see separate spans like `cache request: ...` and `load cache: ...`.
	* We no longer show image pull/extract progress, so no more fancy 2d progress bars.
	* We no longer send the LLB op protobuf in the telemetry.
	* All of these can probably be upstream PRs.

## Notable tweaks

* DagQL no longer marks module functions as 'impure', which was causing parallel calls to never be deduped, leading to duplicate spans.
	* Technically we already cache module functions within a session, and the DagQL server cache has the same lifetime as the session, so this should not be a behavior change, but we'll want to change this if/when we take a swing at module function caching and/or publishing recipes.
* `dagger terminal` no longer does the insane vterm emulation thing, which was originally done so we can support showing progress alongside your active terminal session.
	* Now you should just use `dagger watch` in a separate terminal. Implementing a terminal emulator is not our core competency. It was fun, but also endless toil. The new implementation is _much_ simpler.
* OpenTelemetry is very dependent on `ctx` propagation, and `context.Background()` breaks that, so now we use the new `context.WithoutCancel()` API which is nearly always what we wanted instead, since it keeps all the juicy values but just drops the cancel-propagation aspect.
* We now initialize the TUI frontend _super_ early on - literally before we even initialize telemetry. This is great for various reasons, but the trade-off is we need to do custom barebones flag parsing to handle `-v`/`--debug` and other global flags. We can't defer that to the usual flag parsing since that's tied to module loading and stuff (for e.g. `dagger call`).
* We currently have our own hokey OpenTelemetry Go SDK for logging under `telemetry/env` and `telemetry/sdklog` which should be nuked as soon as the official SDK lands (https://github.com/open-telemetry/opentelemetry-go/pull/4955).
* Deleted various stale commands under `cmd/` that weren't worth maintaining.
	* `otel-collector` (this is just native now)
	* `dagger-graph` (this was coupled to Progrock/Buildkit progress stream)
	* `upload-journal` (replaced by OTLP)

### CODE REVIEW

**Review by vito — COMMENTED** [review]
Good to go now I think, updated the PR description but it's still very rough, ran out of steam and just copy-pasted my notes for now. Also left some inline review comments for subtle things.

Sorry this is such a dense PR, it honestly just requires touching a ton of things. :sweat_smile: 

There's no time pressure to merge this now, in fact we don't want to merge until the Cloud UI is done and shipped. Wanted to get it in a reviewable state ASAP to leave enough time.

**Review by sipsma — COMMENTED** [review]
Just did some initial perusing, looks pretty cool, nice that a lot of the actual usage around writing statuses/events is a lot simpler.

> we no longer show 'CACHED' statuses for vertices

Is that because Buildkit only exposes that via its progress stream? If so, I think we could probably make an upstream change to support `Solve` calls returning whether they were cached or not (or equivalent behavior). Doesn't feel like too monumental a task.

I guess the fact that you could be solving a lot of ops that may be in a mixed state of cached or not might be the hard part, may need to return a map of llb digest to cached state or something (thinking out loud)

**Comment by vito** [comment]
> Is that because Buildkit only exposes that via its progress stream? If so, I think we could probably make an upstream change to support `Solve` calls returning whether they were cached or not (or equivalent behavior). Doesn't feel like too monumental a task.

Yeah, I think there are a few things we should consider PRing upstream:

* Changing how the cached spans work; instead of having separate `cache request:` and `read cache:` spans it seems like we'd just want an event on the original span saying "hey this was cached." Or maybe that's an "and" and not and "or" - TBD, not sure if those other spans serve some purpose I'm not aware of.
* Emitting image pull/extract progress, also as events on individual spans, so we can render those nice 2-dimensional progress bars again.
* Sending exec stdout/stderr over OpenTelemetry logging, though this makes more sense to do after OpenTelemetry's Go SDK supports it.

In general this shines a light on a double-edged sword to OpenTelemetry: it's awesome that we're able to surface all this detail, and even support user-land telemetry in things they run, but it's not so awesome that it won't always align with our DX interests. Just something to watch out for going forward.

File: .gitmodules
Comment (vito) [reply .gitmodules:1]: Not temporary, but not very load-bearing either. We have a custom gRPC service implementing pub/sub for OpenTelemetry, so now we have a dependency on their protobuf definitions. We'll just need to make sure we initialize submodules before we run `go generate ./telemetry` in CI and such. (Which we should probably be doing anyway.)

Alternatively we can look into tools like [Buf](https://buf.build/) which might provide an alternative to the submodule.

File: cmd/dagger/engine.go
Comment (vito) [reply cmd/dagger/engine.go:65]: Yeah, man, honestly it's very dense but I can't say it's without merit; instrumenting the world in a way that's compatible with all these use cases is just very complicated, and even with all the existing complexity we've had to do funky things of our own, like having a "live span processor" that accepts in-flight spans and sends them to the exporter while they're running, whereas normally spans are only exported when they complete.

For me it's mostly cobbled together knowledge + a long ongoing chat with ChatGPT, but maybe the [Glossary](https://opentelemetry.io/docs/concepts/glossary/) would help?

tl;dr for some terms you can expect to see:

* resource: roughly represents a component, e.g. the running process; has typed attributes, usually just a handful of these
* span: probably familiar already but including for completeness; measures a span of time spent doing some task, associated to a resource (analogous to a component), can have arbitrary typed attributes and events, which are all held in memory (so don't use those for e.g. logs)
* log record: a chunk of output, but it's actually a typed value, and also has attributes, is tied to a resource, etc.
* [log|span] processor: accepts logs and spans as they appear and passes them along to an exporter
* [log|spab] exporter: sends a batch of spans to a consumer, e.g. an OTLP endpoint, our CLI frontend, or Cloud
* gRPC unary interceptor: instruments simple gRPC call/response APIs with OpenTelemetry, has both client and server hooks
* gRPC stream interceptor: instruments streaming gRPC call/response APIs with OpenTelemetry, has both client and server hooks
* TracerProvider: basically a factory for Tracers, typically globally configured, since it's what has exporters/span processors configured inside and you don't want to have to DI those everywhere
* Tracer: a factory for spans, initialized with a component name (I've made this align roughly with Go packages, e.g. `dagger.io/core`, which isn't a real Go package, but it hasn't mattered yet)

Drawing the line there for now. :sweat_smile: Feel free to ask anythign else.

And just to reiterate since we're doing something weird: our "catch" is that our CLI and Cloud accept "live" spans, whereas most OpenTelemetry systems (e.g. Honeycomb) only accept completed spans, and will get very confused if you send them "live" spans.

File: cmd/dagger/module.go
Comment (vito) [reply cmd/dagger/module.go:834]: Ah good catch, didn't see it was unused! Removed

**Comment by aluzzardi** [comment]
> Changing how the cached spans work; instead of having separate cache request: and read cache: spans it seems like we'd just want an event on the original span saying "hey this was cached." Or maybe that's "in addition to" and not "instead" - not sure if those other spans serve some purpose I'm not aware of.

👍 

Event AND / OR an attribute would work too -- span attributes can be changed at any point until `End()` is called. The advantage of an attribute is it makes it easy to query/filter/process by that status.

My first instinct would be to have a span attribute to indicate cache status as a quick solid marker, PLUS events detailing the cache process -- retrieval (from where?), unpacking, whatnot.

File: cmd/dagger/engine.go
Comment (aluzzardi) [reply cmd/dagger/engine.go:65]: Yeah, it's pretty cobbled together. I guess for every day development, the tracer/provider/factories yadda yadda don't really matter.

The gist of it is OTEL is actually 3 projects:
- Tracing: which we use extensively
- Logs: which are pretty simple compared to Tracing, and are still experimental
- Metrics: which we are NOT using yet -- but this could be super interesting. @sipsma For instance, the engine logs you added to dump memory/cpu/etc could now use OTEL metrics. We could also push this even further by exporting container-level metrics and graphing that in Cloud

For tracing, the gist of the terminology is here: https://opentelemetry.io/docs/specs/otel/overview/#traces

Translation manual:

- `Trace` == `Run`. A Trace is just a `TraceID` used to tie `Span`s together, exactly like `Run`/`RunID` and vertices.
- `Span` == `Vertex-ish`. It's got a start time, end time, state, and arbitrary key/values attached. It can also have "events" (message + key/value at a point in time during the span) -- I guess similar to `VertexStatuses`?
- `Resource` == `Label` (?). Key/value map attached to all spans. Conventionally in tracing this contains stuff like service name, PID, etc. We are stashing the "engine labels" in there (git information, etc)

**Comment by vito** [comment]
Status update:

* I'd like to merge this week, after tomorrow's release, since after it's merged we can't release until Cloud's new OpenTelemetry visualization ships.
* [ ] Our shim should expose the OTLP endpoints over TCP on `localhost` instead, since some languages/libraries don't support `unix://`. Will try to get to this today, but we can also do it in a second pass.
* The `withExec` API should probably support disabling telemetry, but I'm inclined to hold off on this until we ship proper/full OpenTelemetry integration, rather than the current half-baked version that's just coming along for the ride from Buildkit and causing confusion.
* The PR description doesn't cover this yet, but each SDK SHOULD implement OpenTelemetry support too. This PR covers the Go SDK, but really it's another thing each SDK should have in scope to be "complete" (on top of interfaces, modules, etc.)
  * Full support means detecting `$TRACEPARENT`, propagating the span context through to HTTP headers to the engine, and allowing the user to create spans of their own.
    * For legacy non-module SDK usage, `$TRACEPARENT` will be set by `dagger run` and needs to be captured. 
    * For module SDK usage, `$TRACEPARENT` and `OTEL_*` will be set by the shim and should be automatically wired up by the SDK runtime. For example, when a Go Zenith function gets called, its `context.Context` should be already instrumented.
    * The runtime SHOULD send spans "live". It's kind of a pain to support that in every language, since it's not a native OpenTelemetry feature and requires custom code. It will degrade gracefully if not implemented (users just won't see custom spans until they complete), so this is "extra credit".
  * It's sad that there's more work on SDK authors, but it's nice that it's an open standard now, so it should be pretty vanilla to implement (except for "live" spans), and it gives users pretty cool new capabilities (e.g. emitting a span for each test in our integration test suite would be really cool).

**Comment by gerhard** [comment]
Putting this on your radar @wingyplus @kjuulh @zeroware @jcsirot since it will require some work on the SDK side. Once we finish Typescript & Python (Go is already done part of this) it might be easier to replicate the functionality.

**Comment by kjuulh** [comment]
> Putting this on your radar @wingyplus @kjuulh @zeroware @jcsirot since it will require some work on the SDK side. Once we finish Typescript & Python (Go is already done part of this) it might be easier to replicate the functionality.

Thanks for the heads up @gerhard

**Comment by vito** [comment]
Just a quick note that the Go SDK part isn't fully complete yet; I still need to finish the Go module runtime integration so that it automatically instruments the `context.Context` passed to functions. Working on it now, but bogged down in annoying codegen <-> SDK library code sharing trickery.

**Comment by wingyplus** [comment]
> Putting this on your radar @wingyplus @kjuulh @zeroware @jcsirot since it will require some work on the SDK side. Once we finish Typescript & Python (Go is already done part of this PR) it might be easier to replicate the functionality.

Thanks for the heads up @gerhard

**Comment by wingyplus** [comment]
This looks really interesting but my brain is nearly shutdown now. Let me take a look in details tomorrow. @__@

File: core/query.go
Comment (sipsma) [inline core/query.go:79]: Looking through the buildkit code, I'm a bit concerned that if we don't read its progress it will still end up writing it (including exec-op stdout/stderr) and then caching that all in memory for the duration of the job since it's never read.

I am not 100% sure that's actually what happens, but may be worth a test with a container that writes a ton to stdio and check whether the memory explodes. If it does, I guess we could leave in a reader for the progress that just discards it (for now anyways, obviously would be nice if we could modify upstream to just disable writing any of this)

File: dagql/call/id.go
Comment (sipsma) [inline dagql/call/id.go:62]: Just FYI the reason I didn't add this originally is that it allows mutation of the underlying pb which would break the digest. But I'm fine with it in place as long as the callers pinky promise to not mutate it 😄 (may be worth a comment on the function)

**Review by sipsma — COMMENTED** [review]
Overall LGTM on another pass, will mostly need to rely on a lot of testing and looking at the progress output to verify since there's so much new stuff (not a complaint, just how it has to be). Just left a couple comments and saw commented out chunks of code in a few places that should be cleaned up, otherwise good to ship I think.

File: cmd/dagger/main.go
Comment (vito) [reply cmd/dagger/main.go:172]: Yeah, this part is pretty annoying.

What we want is that for `dagger call -vvv -m github.com/foo/bar fizz --buzz 123` we can see the `-vvv` part and apply that change to the frontend.

We can't use the regular flag parsing stack because that all happens in `Execute` which will start initializing modules and stuff before we can even initialize the frontend and tell it to display with a verbosity level of 3.

And `pflags` doesn't expose a way to ignore extra flags; if you try to parse the above with a smaller set of global flags it'll error because it doesn't know the `--buzz` flag.

Also, if we actually register a flag called `"verbose"` that will conflict with a module-provided `verbose` flag. This is technically true for any module-provided flag whose name collides with a global flag, I just happened to run into this while testing certain modules, and we didn't have a `--verbose` flag before. The only way to really avoid this is either detecting conflicts and renaming them (kinda gross), or _only_ use short flags for these 'global things', which  we can _almost_ do, except `pflag` doesn't let you configure _only_ short flags - you always have to register a long flag.

I considered calling the long flags something stupid like `--please-verbose` but at that point recognized my desperation and went back to the comfort of this much smaller, slightly more comfortable rat's nest that is this `for {}` loop. :sob:

File: dagql/call/id.go
Comment (wingyplus) [inline dagql/call/id.go:63]: 😂

File: cmd/codegen/generator/go/templates/src/_dagger.gen.go/defs.go.tmpl
Comment (wingyplus) [inline cmd/codegen/generator/go/templates/src/_dagger.gen.go/defs.go.tmpl:170]: The `$TRACEPARENT` injection works only `dagger run` (or dagger call also, I'm not sure), right? Just thinking about auto-provision (automatic download the dagger binary), but that case looks impossible to set it.

Comment (vito) [reply cmd/codegen/generator/go/templates/src/_dagger.gen.go/defs.go.tmpl:170]: You're right, there's not much we can do when it's not wrapped in `dagger call` or `dagge run` since there's no opportunity to establish a root span - unless the user's code is doing that themselves. I think that's OK.

The comment says `dagger run` but this will also be the case when we're calling a module function, since the `shim` executable sets `$TRACEPARENT` there too. Will update the comment.

**Comment by helderco** [comment]
Pushed Python support. Simple module with a function that creates a new span:
```python
from opentelemetry import trace

from dagger import dag, function, object_type

tracer = trace.get_tracer(__name__)


@object_type
class Woof:
    @function
    async def echo(self, msg: str) -> str:
        return await (
            dag.container()
            .from_("alpine:latest")
            .with_exec(["echo", msg])
            .stdout()
        )

    @function
    async def span(self) -> str:
        with tracer.start_span("parrot"):
            return await self.echo("🚀")
```

**Comment by helderco** [comment]
> * The runtime SHOULD send spans "live". It's kind of a pain to support that in every language, since it's not a native OpenTelemetry feature and requires custom code. It will degrade gracefully if not implemented (users just won't see custom spans until they complete), so this is "extra credit".

@vito, does the Go SDK have this now? Seems like the code handling this is used only in core/engine. Same question for logs support.

**Comment by vito** [comment]
@helderco Not yet - that's the part that's been giving me trouble.

There are two code trees that help with Go OpenTelemetry:

* `telemetry/inflight/`
* `telemetry/sdklog/`

For the Go runtime I would need to vendor these in a way similar to the `querybuilder`, but that technique isn't so easy for the above two packages because they depend on sibling packages, and so they has import statements that need to be modified (or redirected with `go replace` I suppose) to point to the codegen'd path.

I think I can move `telemetry/inflight/` over somewhat easily, since its only peer dependency is `telemetry/env/` and that can just be inlined.

I can't move `telemetry/sdklog/` as easily because it's a lot more code with a lot more imports. I can just skip this for now for the Go runtime, since the module can just log to stdout/stderr. But it came up because I _do_ need it for the shim, and I was trying to have them all share the same code, so that led me to a dead end.

Fortunately this `sdklog` package will be going away anyway, once the OpenTelemetry Go SDK support is done (https://github.com/open-telemetry/opentelemetry-go/pull/4955).

In any case, I'm also wondering if there's a less painful long-term solution I can find for sharing code between the Go SDK, engine, and Go runtime - like maybe an automated `go replace` rule for `dagger.io/dagger` since modules probably shouldn't have a dependency on a particular Go SDK version anyway. But I've been busy fixing up other things, so haven't gotten back to it yet.

File: cmd/dagger/main.go
Comment (vito) [reply cmd/dagger/main.go]: Ok, phew, found a less janky solution to this: https://github.com/dagger/dagger/pull/6835/commits/d95d6641b7e5e43fcefa77bc8abe0972f42abcc5

**Comment by vito** [comment]
Update: ok, Go SDK support for custom spans is implemented: https://github.com/dagger/dagger/pull/6835/commits/31b303b01d7877003aec67d54ccb260df67cbb8c

But honestly it's not a great reference for other SDK authors; it's a bit hairy for not great reasons. I'd check out the Python SDK (or the Elixir SDK :tada: - thanks @wingyplus!) instead. You can totally reference the 'live span' code though, that part's just copy-pasta from `telemetry/inflight/` and shows how to implement it.

@sipsma @jedevc WDYT of refactoring the Go SDK codegen <=> Go SDK runtime <=> Dagger engine code sharing?[^1] Rough idea is to change the runtime to work via `replace dagger.io/dagger => ./internal/dagger`, the same way the engine works. That way we can share code more easily without worrying about import paths. (Motivation: https://github.com/dagger/dagger/pull/6835#issuecomment-2025408579)

That would result in the easiest central place to share code being the Go SDK, which kind of tracks I guess, since we're already incentivized to keep its dependencies minimal, as we would with any code shared between all three.

Would save that for a future PR, this one's already way too big.

Also @sipsma I think all the feedback is addressed now, modulo recent changes. I didn't find time to double-check the Buildkit memory thing so I just made it send to `io.Discard` and tweaked the comment. But the flag parsing situation is way less gross now.

:pray: 

[^1]: Having a hard time figuring out what to call these, lemme know if there are better terms. By "Go SDK codegen" I mean `dagger.io/dagger`, by "Go SDK runtime" I mean the module codegen'd package, and by the Dagger engine I mean... the Dagger engine.

**Comment by wingyplus** [comment]
I found an issue when I'm working on #6962. I create a file to `sdk/elixir/test.exs` with the snippet below

```elixir
Application.ensure_all_started(:inets)

Mix.install([{:dagger, path: "."}])

defmodule Test do
  def run(dag \\ Dagger.connect!()) do
    dag
    |> Dagger.Client.container()
    |> Dagger.Container.from("alpine")
    |> Dagger.Container.with_exec(~w"echo hello")
  |> Dagger.Sync.sync()
  after
    Dagger.close(dag)
  end
end

Test.run
```

The script cannot compile because I'm accidentally make the sdk compilation error. When run with `with-dev dagger run elixir test.exs` I found this error

```
wingyplus@WINGYMOMO:~/src/github.com/dagger/dagger-elixir-otel/sdk/elixir$ with-dev dagger run elixir test.exs
==> dagger
Compiling 2 files (.ex)

== Compilation error in file lib/dagger/core/graphql_client/httpc.ex ==
** (SyntaxError) invalid syntax found on lib/dagger/core/graphql_client/httpc.ex:23:7:
    error: syntax error before: '['
    │
 23 │       []
    │       ^
    │
    └─ lib/dagger/core/graphql_client/httpc.ex:23:7
    (elixir 1.16.2) lib/kernel/parallel_compiler.ex:428: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/8
could not compile dependency :dagger, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile dagger --force", update it with "mi
x deps.update dagger" or clean it with "mix deps.clean dagger"
21:42:58 WRN failed to get repo HEAD err="reference not found"
exit status 1
wingyplus@WINGYMOMO:~/src/github.com/dagger/dagger-elixir-otel/sdk/elixir$
```

Let's ignore Elixir error for now. The issue is when error occurred, my shell display is now broken for some reason. :(

**Review by sipsma — APPROVED** [review]

**Comment by vito** [comment]
Thanks for the approval. :pray: 

Integration tests were failing for the Python SDK, related to the new `opentelemetry` dependency, so I've just reverted those for now so we can get this PR merged first. Each SDK can just be a separate PR.

Will merge once checks are green.

**Comment by vito** [comment]
Update: some Python related tests flaked due to a concurrent map write here:

https://github.com/dagger/dagger/blob/main/sdk/python/runtime/discovery.go#L250

This is unrelated, but I've piled a quick fix onto my PR (just added a mutex), since it failed twice in a row.

**Pull request merged**
